File: backend/src/main/java/com/nischal/backend/service/MenuService.java
package com.nischal.backend.service;

import com.nischal.backend.dto.menu.CategoryRequest;
import com.nischal.backend.dto.menu.CategoryResponse;
import com.nischal.backend.dto.menu.MenuItemRequest;
import com.nischal.backend.dto.menu.MenuItemResponse;

import java.util.List;

public interface MenuService {

    // Categories
    List<CategoryResponse> getAllCategories();
    CategoryResponse getCategoryById(Long id);
    CategoryResponse createCategory(CategoryRequest request);
    CategoryResponse updateCategory(Long id, CategoryRequest request);
    void deleteCategory(Long id);

    // Menu Items
    List<MenuItemResponse> getAllMenuItems();
    List<MenuItemResponse> getMenuItemsByCategory(Long categoryId);
    MenuItemResponse getMenuItemById(Long id);
    MenuItemResponse createMenuItem(MenuItemRequest request);
    MenuItemResponse updateMenuItem(Long id, MenuItemRequest request);
    void deleteMenuItem(Long id);
    List<MenuItemResponse> searchMenuItems(String keyword);
    List<MenuItemResponse> getFullMenuGroupedByCategory();
}

File: frontend/src/services/menu.service.ts
import { apiClient } from './api.service';
import type {
  Category,
  MenuItem,
  CategoryRequest,
  MenuItemRequest,
  TableInfo,
} from '../types/menu.types';

export const menuService = {
  // Public (no auth required)

  async getFullMenu(): Promise<MenuItem[]> {
    return apiClient.get<MenuItem[]>('/menu');
  },

  async getCategories(): Promise<Category[]> {
    return apiClient.get<Category[]>('/menu/categories');
  },

  async getItemsByCategory(categoryId: number): Promise<MenuItem[]> {
    return apiClient.get<MenuItem[]>(`/menu/categories/${categoryId}/items`);
  },

  async getMenuItemById(id: number): Promise<MenuItem> {
    return apiClient.get<MenuItem>(`/menu/items/${id}`);
  },

  async searchMenu(keyword: string): Promise<MenuItem[]> {
    return apiClient.get<MenuItem[]>(`/menu/search?keyword=${encodeURIComponent(keyword)}`);
  },

  async resolveQrToken(qrToken: string): Promise<TableInfo> {
    return apiClient.get<TableInfo>(`/qr/resolve/${qrToken}`);
  },

  // Admin / Owner operations (auth required)

  async createCategory(data: CategoryRequest): Promise<Category> {
    return apiClient.post<Category>('/menu/categories', data);
  },

  async updateCategory(id: number, data: CategoryRequest): Promise<Category> {
    return apiClient.put<Category>(`/menu/categories/${id}`, data);
  },

  async deleteCategory(id: number): Promise<void> {
    return apiClient.delete<void>(`/menu/categories/${id}`);
  },

  async createMenuItem(data: MenuItemRequest): Promise<MenuItem> {
    return apiClient.post<MenuItem>('/menu/items', data);
  },

  async updateMenuItem(id: number, data: MenuItemRequest): Promise<MenuItem> {
    return apiClient.put<MenuItem>(`/menu/items/${id}`, data);
  },

  async deleteMenuItem(id: number): Promise<void> {
    return apiClient.delete<void>(`/menu/items/${id}`);
  },

  getQrCodeUrl(tableId: number): string {
    return `http://localhost:8081/api/v1/qr/table/${tableId}`;
  },
};
